pkg/controller/replicaset/replica_set.go

```diff
@@ -681,18 +722,65 @@ func slowStartBatch(count int, initialBatchSize int, fn func() error) (int, erro
 	return successes, nil
 }
 
-func getPodsToDelete(filteredPods []*v1.Pod, diff int) []*v1.Pod {
+// getIndirectlyRelatedPods returns all pods that are owned by any ReplicaSet
+// that is owned by the given ReplicaSet's owner.
+func (rsc *ReplicaSetController) getIndirectlyRelatedPods(rs *apps.ReplicaSet) ([]*v1.Pod, error) {
+	var relatedPods []*v1.Pod
+	seen := make(map[types.UID]*apps.ReplicaSet)
+	for _, relatedRS := range rsc.getReplicaSetsWithSameController(rs) {
+		selector, err := metav1.LabelSelectorAsSelector(relatedRS.Spec.Selector)
+		if err != nil {
+			return nil, err
+		}
+		pods, err := rsc.podLister.Pods(relatedRS.Namespace).List(selector)
```

Comment:
How expensive is this in very large namespaces with either lots of RS or lots of pods under those RS?

Reply:
Does this make a replica set of size 15k potentially extremely expensive to run?  Or if we only had one, it would be fine?

Reply:
I hope that the use of informers for pods and replica sets will mitigate performance problems, but I have not tested at scale.  Can you recommend a framework, or should I do some rudimentary instrumentation of the controller and perform some manual testing?

Reply:
It's not the cheapest operation, I agree. But I think we're not regressing much here.
Something similar is already happening in deployment controller:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L536

It's important that we operate on pointers here (which are shared with cache), so we don't do that many memory allocations.
So while it would be better to avoid this, I can probably live with that if the feature is important enough (reliability of rollout seems worth it, though personally I'm not convinced that this will really help much).